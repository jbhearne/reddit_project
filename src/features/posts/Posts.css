
ul {
    list-style: none;
    padding: 0;
}


.posts_container {
    display: block;
    columns: 2 200px;
    column-gap: 2rem;
    margin: 10px 30px 10px 10px;
}

@keyframes growContainer {
    from {
        width: 0%;
        height: 0%;
    }

    to {
        width: 100%;
        height: 100%;
    }
}


@media only screen and (max-width: 700px) {
    .posts_container {
        display: block;
        columns: 1 200px;
        column-gap: 2rem;
        margin: 10px 30px 10px 10px;
    }
}

@media only screen and (min-width: 1400px) {
    .posts_container {
        display: block;
        columns: 3 200px;
        column-gap: 2rem;
        margin: 10px 30px 10px 10px;
    }
}

.post_box {
    border: solid #34a2c3 6px;
    margin-bottom: 1rem;
    display: inline-block;
    width: 100%;
    background-color: white;
    border-radius: 10px;
    padding: .25rem;
    animation-name: growContainer;
    animation-duration: 1s;
}

.post_box video {
    width: 90%;
}

.post_box img {
    width: 90%;
}

.post_box a {
    text-decoration: none;
    color: #b41e75;
}

.post_box .title {
    width: 100%;
    font-size: 1rem;
}

.post_box .author {
    display: inline-block;
    width: 40%;
    text-align: left;
    margin-left: 5%;
    font-size: .8rem;
    color: #777aab;
}

.post_box .date_upvotes {
    display: inline-block;
    width: 50%;
    text-align: right;
    margin-right: 5%;
    font-size: .7rem;
    font-weight: 400;
    color: #8ebd79;
}