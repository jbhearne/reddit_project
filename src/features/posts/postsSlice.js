import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { fetchReddit } from '../../app/redditAPI';
import { convertUnixTime } from '../../app/convertUnixTime';

export const fetchPosts = createAsyncThunk(
    'posts/fetchPosts',
    async (url) => {
        const response = await fetchReddit(url);
        console.log(response.children[0].data.gallery_data)
        //const res = response
        //console.log(res)

        const mapp = response.children.map(child => {
            console.log(child)
            
            const getImageUrls = () => {
                let imageUrls = []
                
                if (child.data.gallery_data) {
                    const imageIdData = child.data.gallery_data.items.filter(idData => child.data.media_metadata[idData.media_id].e === 'Image')
                    imageUrls = imageIdData.map(image => {
                        return child.data.media_metadata[image.media_id].p[0].u.replace('https://preview.redd.it', 'https://i.redd.it');
                    })
                } else if (child.data.post_hint === 'image') {
                    imageUrls.push(child.data.url);
                };
                return imageUrls
            }

            const getMedia = () => {
                const media = {}
                if (child.data.media) {
                    if (child.data.media.oembed) {
                        media.embed = child.data.media.oembed.html.replace(/&lt;/g, '<').replace(/&gt;/g, '>')
                    } else if (child.data.media.reddit_video) {
                        media.video = child.data.media.reddit_video.fallback_url
                    }
                }
                return media
            }

            return {
                author: child.data.author,
                title: child.data.title,
                date: convertUnixTime(child.data.created),
                upvotes: child.data.score,
                text: child.data.selftext,
                images:  getImageUrls(),
                media:  getMedia(),
                url: child.data.url,
                id: child.data.id
            }
        })
        //console.log(response)
        //console.log(res)
        console.log(mapp[0])
        return mapp
    }
)

export const postsSlice = createSlice({
    name: 'posts',
    initialState: {
        prefix: 'https://www.reddit.com',
        path: '',
        url: '',
        posts: [],
        isLoading: false,
        hasError: false
    },
    reducers: {
        setPostsPath: (state, action) => {
            state.path = action.payload
        },
        setPostsUrl: (state, action) => {
            state.url = state.prefix + state.path;
        },
        setPostsPrefix: (state, action) => {
            state.prefix = action.payload;
        }
    },
    extraReducers: {
        [fetchPosts.pending]: (state, action) => {
            state.isLoading = true;
            state.hasError = false;
        },
        [fetchPosts.fulfilled]: (state, action) => {
            state.posts = action.payload
            state.isLoading = false;
            state.hasError = false;
        },
        [fetchPosts.rejected]: (state, action) => {
            state.isLoading = false;
            state.hasError = true;
        }
    }
})

export const selectPosts = (state) => state.posts.posts;
export const selectPostsUrl = (state) => state.posts.url;
export const selectPostsPath = (state) => state.posts.path;
export const selectPostsPrefix = (state) => state.posts.prefix;
export const setPostsUrl = postsSlice.actions.setPostsUrl;
export const setPostsPrefix = postsSlice.actions.setPostsPrefix;
export const setPostsPath = postsSlice.actions.setPostsPath;

export default postsSlice.reducer