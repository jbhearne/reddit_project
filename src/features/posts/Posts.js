import React, { useState, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Post } from './post/Post'
import { selectPosts, fetchPosts, selectUrl } from './postsSlice';
import './Posts.css'
import {}

export function Posts() {

    const posts = useSelector(selectPosts);
    const url = useSelector(selectUrl);
    const dispatch = useDispatch();

    useEffect(() => {
        dispatch(fetchPosts(url))
    }, [url])

    return (
        <ul className='posts_container'>
            {posts.map(post => {
                return (
                    <li className='post_box' key={post.id}><Post post={post} /></li>
                )
            })}
        </ul>
    )
}