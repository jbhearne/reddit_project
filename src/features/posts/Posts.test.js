import { useSelector, useDispatch } from "react-redux";
import { selectPosts, selectPostsUrl, selectPostsPath, selectPostsPrefix, fetchPosts } from './postsSlice';
import { Posts } from './Posts'
import { fakePostsComponentData } from '../../../__tests__/fakeComponent.data'
import renderer from 'react-test-renderer'
import { BrowserRouter as Router } from 'react-router-dom';

jest.mock('react-redux')
jest.mock('./postsSlice')

describe('Posts component', () => {
    let mockComponentPosts = fakePostsComponentData

    it('should render', () => {
        fetchPosts.mockImplementation(url => {
            mockComponentPosts = fakePostsComponentData
        })
        
        useSelector.mockImplementation(selection => {
            if (selection === selectPosts) {
                return mockComponentPosts
            } else if (selection === selectPostsUrl) {
                return 'URLtest'
            } else {
                return 'nope'
            }
        })
        
        useDispatch.mockReturnValue((dis) => dis)

        const component = renderer.create(
            <Router>
                <Posts />
            </Router>
        )
        let tree = component.toJSON()
        expect(tree).toMatchSnapshot();
        tree = component.toJSON()
        expect(tree).toMatchSnapshot();

    })

})