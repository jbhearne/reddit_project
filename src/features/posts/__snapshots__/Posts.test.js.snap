// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts component should render 1`] = `
<ul
  className="posts_container"
>
  <li
    className="post_box"
  >
    <article
      className="post"
    >
      <h3
        className="title"
      >
        <a
          href="testUrl"
          target="_blank"
        >
          testTitle
        </a>
      </h3>
      <h4
        className="author"
      >
        testAuthor
      </h4>
      <h5
        className="date_upvotes"
      >
        testDate - Upvotes: 1
      </h5>
      <p>
        testText
      </p>
      <div>
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "testMedia",
            }
          }
        />
      </div>
      <ul>
        <li>
          <img
            src="testImage11"
          />
        </li>
        <li>
          <img
            src="testImage12"
          />
        </li>
      </ul>
    </article>
  </li>
  <li
    className="post_box"
  >
    <article
      className="post"
    >
      <h3
        className="title"
      >
        <a
          href="testUrl2"
          target="_blank"
        >
          testTitle2
        </a>
      </h3>
      <h4
        className="author"
      >
        testAuthor2
      </h4>
      <h5
        className="date_upvotes"
      >
        testDate2 - Upvotes: 12
      </h5>
      <p>
        testText2
      </p>
      <div>
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "testMedia2",
            }
          }
        />
      </div>
      <ul>
        <li>
          <img
            src="testImage21"
          />
        </li>
        <li>
          <img
            src="testImage22"
          />
        </li>
      </ul>
    </article>
  </li>
  <li
    className="post_box"
  >
    <article
      className="post"
    >
      <h3
        className="title"
      >
        <a
          href="testUrl3"
          target="_blank"
        >
          testTitle3
        </a>
      </h3>
      <h4
        className="author"
      >
        testAuthor3
      </h4>
      <h5
        className="date_upvotes"
      >
        testDate3 - Upvotes: 13
      </h5>
      <p>
        testText3
      </p>
      <div>
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "testMedia3",
            }
          }
        />
      </div>
      <ul>
        <li>
          <img
            src="testImage31"
          />
        </li>
        <li>
          <img
            src="testImage32"
          />
        </li>
      </ul>
    </article>
  </li>
</ul>
`;

exports[`Posts component should render 2`] = `
<ul
  className="posts_container"
>
  <li
    className="post_box"
  >
    <article
      className="post"
    >
      <h3
        className="title"
      >
        <a
          href="testUrl"
          target="_blank"
        >
          testTitle
        </a>
      </h3>
      <h4
        className="author"
      >
        testAuthor
      </h4>
      <h5
        className="date_upvotes"
      >
        testDate - Upvotes: 1
      </h5>
      <p>
        testText
      </p>
      <div>
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "testMedia",
            }
          }
        />
      </div>
      <ul>
        <li>
          <img
            src="testImage11"
          />
        </li>
        <li>
          <img
            src="testImage12"
          />
        </li>
      </ul>
    </article>
  </li>
  <li
    className="post_box"
  >
    <article
      className="post"
    >
      <h3
        className="title"
      >
        <a
          href="testUrl2"
          target="_blank"
        >
          testTitle2
        </a>
      </h3>
      <h4
        className="author"
      >
        testAuthor2
      </h4>
      <h5
        className="date_upvotes"
      >
        testDate2 - Upvotes: 12
      </h5>
      <p>
        testText2
      </p>
      <div>
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "testMedia2",
            }
          }
        />
      </div>
      <ul>
        <li>
          <img
            src="testImage21"
          />
        </li>
        <li>
          <img
            src="testImage22"
          />
        </li>
      </ul>
    </article>
  </li>
  <li
    className="post_box"
  >
    <article
      className="post"
    >
      <h3
        className="title"
      >
        <a
          href="testUrl3"
          target="_blank"
        >
          testTitle3
        </a>
      </h3>
      <h4
        className="author"
      >
        testAuthor3
      </h4>
      <h5
        className="date_upvotes"
      >
        testDate3 - Upvotes: 13
      </h5>
      <p>
        testText3
      </p>
      <div>
        <div
          dangerouslySetInnerHTML={
            Object {
              "__html": "testMedia3",
            }
          }
        />
      </div>
      <ul>
        <li>
          <img
            src="testImage31"
          />
        </li>
        <li>
          <img
            src="testImage32"
          />
        </li>
      </ul>
    </article>
  </li>
</ul>
`;
